src/ngraph/builder/reshape.cpp

```diff
@@ -201,20 +204,97 @@ shared_ptr<Node> builder::opset1::transpose(const Output<Node>& value)
     return builder::opset1::reorder_axes(value, axes_order);
 }
 
-shared_ptr<Node> builder::opset1::flatten(const Output<Node>& value, int axis)
+namespace ngraph
 {
-    auto data_shape = value.get_shape();
-
-    // First dimension of output tensor is the product of [d_0, ... d_{axis-1}] dimensions of input
-    // tensor. The last dimension is the product of the rest of input tensor dimensions:
-    // [d_{axis}, ..., d_n]
-    size_t first_dim_size =
-        accumulate(begin(data_shape), next(begin(data_shape), axis), 1UL, multiplies<size_t>());
+    namespace builder
+    {
+        namespace opset1
+        {
+            namespace
+            {
+                ///
+                /// \brief      Return the node representing normalized axis with respect to
+                ///             provided rank.
+                ///
+                /// \param[in]  node_rank  The node representing rank used for normalization.
+                /// \param[in]  axis       The axis value to be normalized.
+                ///
+                /// \return     The new Constant node representing normalized axis value.
+                ///
+                std::shared_ptr<Node>
+                    get_normalized_axis_node(const std::shared_ptr<Node> node_rank, int64_t axis)
+                {
+                    auto axis_node =
+                        ngraph::opset1::Constant::create(element::i64, Shape{1}, {axis});
+                    // shortcut for alredy positive value
+                    if (axis >= 0)
+                    {
+                        return axis_node;
+                    }
+
+                    // TODO: What if axis value is beyond acceptable values? [-node_rank,
+                    // node_rank-1]
+                    return make_shared<ngraph::opset1::Add>(node_rank, axis_node);
+                }
+            } // opset1
+        }     // builder
+    }         // ngraph
+}
 
-    size_t last_dim_size =
-        accumulate(next(begin(data_shape), axis), end(data_shape), 1UL, multiplies<size_t>());
+shared_ptr<Node> builder::opset1::flatten(const Output<Node>& value, int axis)
+{
+    if (value.get_partial_shape().is_static())
```

Comment:
Please get rid of this branch at all, it breaks reshape-ability of the model (setting new Parameter shapes and propagating them down the function).

Reply:
Unfortunately at this moment removing this branch breaks our current support for ONNX.